Smarter handling if queue contains only "no-signal" ops.

bensander opened the pull request:
Commands can be submitted w/o a completion signal.
Now if queue contains only these "no-signal" ops and isEmpty()
is called, then we add a barrier and return false.  Subsequent
calls to isEmpty() will test the signal associated with the barrier
to see if the queue is empty.

bensander committed:
Smarter handling if queue contains only "no-signal" ops.

Commands can be submitted w/o a completion signal.
Now if queue contains only these "no-signal" ops and isEmpty()
is called, then we add a barrier and return false.  Subsequent
calls to isEmpty() will test the signal associated with the barrier
to see if the queue is empty.

AlexVlx commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1494,22 +1494,31 @@ class HSAQueue final : public KalmarQueue
-            if (asyncOps[i] != nullptr) {
-                auto &asyncOp = asyncOps[i];
+        bool oldestNoSignal=false;
+        for (int i = asyncOps.size()-1; i >= 0;  i--) {
```
@bensander couldn't this be just be find_if with reverse iterators + the associated processing? Which is to say something along the lines of: 
```cpp
const auto oldest = find_if(
    asyncOps.crbegin(), asyncOps.crend(), [](... asyncOp) { return asyncOp != nullptr; });

if (oldest != asyncOps.crend()) { /* do the check for signal */ }
```
The loop is somewhat unnatural as is (it appears more customary to use while loops when counting down).

bensander replied:
yep, I think this can be refactored as you suggest, will submit a new one.  It started as a forward loop ; now that it runs backward all we really have to do is find the oldest op and see if it has a signal or not.

AlexVlx commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1494,22 +1494,31 @@ class HSAQueue final : public KalmarQueue
                 hsa_signal_t signal = *(static_cast <hsa_signal_t*> (asyncOp->getNativeHandle()));
                 if (signal.handle) {
+                    oldestNoSignal=false;
                     hsa_signal_value_t v = hsa_signal_load_relaxed(signal);
```
Are we sure relaxed with no fence is what we want here? I believe that this is concurrently written by the CP, is it not?

bensander replied:
This is unchanged from code before, but probably should be changed.  It comes down to what effect is implied on the memory model if a user submits a command and then checks if the associated accelerator view is empty - probably would assume that the results of that command are visible.

AlexVlx reviewed (changes requested):
In general LGTM, but could you please peek at the 2 items I mentioned? Thank you @bensander.

bensander committed:
Refactor smarter nosignal op

- Use simple find_if loop and process oldest op only.
- Use hsa_signal_load_acquire.
- Bonus: move signal,signalIndex,getNativeHandle() into HSAOp
   (since common across all ops).

bensander committed:
Rename _signalIndex and _signal

AlexVlx commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -4970,14 +4969,14 @@ HSACopy::waitComplete() {
-        if (signal.handle) {
-            hsa_signal_load_acquire(signal);
+        if (_signal.handle) {
+            hsa_signal_load_acquire(_signal);
```
Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

AlexVlx commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -4970,14 +4969,14 @@ HSACopy::waitComplete() {
 
     // Wait on completion signal until the async copy is finished
-    hsa_signal_wait_acquire(signal, HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, waitMode);
+    hsa_signal_wait_acquire(_signal, HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, waitMode);
```
Unless I am mistaken, this is deprecated, with `hsa_signal_wait_scacquire` being the current / preferred form.

AlexVlx commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -4847,7 +4844,7 @@ static std::string fenceToString(int fenceBits)
 
 bool HSABarrier::isReady() override {
-    bool ready = (hsa_signal_load_acquire(signal) == 0);
+    bool ready = (hsa_signal_load_acquire(_signal) == 0);
```
Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

AlexVlx commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -4856,7 +4853,7 @@ bool HSABarrier::isReady() override {
 
 bool HSACopy::isReady() override {
-    bool ready = (hsa_signal_load_acquire(signal) == 0);
+    bool ready = (hsa_signal_load_acquire(_signal) == 0);
```
Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

AlexVlx commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -4699,10 +4696,10 @@ HSABarrier::waitComplete() {
 
     // Wait on completion signal until the barrier is finished
-    hsa_signal_wait_acquire(signal, HSA_SIGNAL_CONDITION_EQ, 0, UINT64_MAX, waitMode);
+    hsa_signal_wait_acquire(_signal, HSA_SIGNAL_CONDITION_EQ, 0, UINT64_MAX, waitMode);
```
Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

AlexVlx commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -4865,7 +4862,7 @@ bool HSACopy::isReady() override {
 
 bool HSADispatch::isReady() override {
-    bool ready = (hsa_signal_load_acquire(signal) == 0);
+    bool ready = (hsa_signal_load_acquire(_signal) == 0);
```
Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

AlexVlx commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -5388,11 +5387,11 @@ HSACopy::syncCopyExt(hc::hcCommandKind copyDir, const hc::AmPointerInfo &srcPtrI
         if (hsa_status == HSA_STATUS_SUCCESS) {
             DBOUT(DB_COPY, "HSACopy::syncCopyExt(), wait for completion...");
-            hsa_signal_wait_relaxed(signal, HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, waitMode);
+            hsa_signal_wait_relaxed(_signal, HSA_SIGNAL_CONDITION_LT, 1, UINT64_MAX, waitMode);
```
Should we put this in a loop, since it's both a relaxed wait and it is legal for a signal to spuriously return?

bensander replied:
Most of these are already present in the code and unrelated to the new functionality, lets treat as a new PR.

AlexVlx commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -5376,10 +5375,10 @@ HSACopy::syncCopyExt(hc::hcCommandKind copyDir, const hc::AmPointerInfo &srcPtrI
+        _signalIndex = ret.second;
 
-        hsa_signal_store_relaxed(signal, 1);
+        hsa_signal_store_relaxed(_signal, 1);
```
Should either consider using screlease or a release fence here, to ensure no reordering / fully observable precursors as far as the CP is concerned?

bensander replied:
These are resetting signal value so relaxed seems fine.     CP will store-release to set to 0 ; any consumers wait for the 0 value so that is important point to release.

AlexVlx commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -4434,11 +4431,11 @@ HSADispatch::waitComplete() {
 
         // wait for completion
-        if (hsa_signal_wait_acquire(signal, HSA_SIGNAL_CONDITION_LT, 1, uint64_t(-1), waitMode)!=0) {
+        if (hsa_signal_wait_acquire(_signal, HSA_SIGNAL_CONDITION_LT, 1, uint64_t(-1), waitMode)!=0) {
```
Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

AlexVlx commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1494,22 +1486,29 @@ class HSAQueue final : public KalmarQueue
+        if (oldest != asyncOps.crend()) {
+            hsa_signal_t signal = *(static_cast <hsa_signal_t*> ((*oldest)->getNativeHandle()));
+            if (signal.handle) {
+                hsa_signal_value_t v = hsa_signal_load_acquire(signal);
```
Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

AlexVlx commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1478,7 +1470,7 @@ class HSAQueue final : public KalmarQueue
                 hsa_signal_t signal = *(static_cast <hsa_signal_t*> (asyncOp->getNativeHandle()));
                 if (signal.handle) {
-                    hsa_signal_value_t v = hsa_signal_load_relaxed(signal);
+                    hsa_signal_value_t v = hsa_signal_load_acquire(signal);
```
Unless I am mistaken, this is deprecated, with `hsa_signal_load_scacquire` being the current / preferred form.

AlexVlx reviewed (commented):
Thank you for the changes, LGTM. Now, since you went beyond the call of duty, and touched a few other bits, there are a few other (possibly pedantic :) ) concerns, but it is your call if you want to address them or not.

bensander committed:
Use hsa_signal_load_scacquire.

scchan commented on lib/hsa/mcwamp_hsa.cpp:
```diff
@@ -1494,22 +1486,29 @@ class HSAQueue final : public KalmarQueue
                 }
+            } else {
+                // oldest has no signal - enqueue a new one:
+                auto marker = EnqueueMarker(hc::system_scope);
```
what's the reason for enqueueing a marker?

bensander replied:
If last command had no completion signal then we have no way to tell if he queue is empty or not.  So we enque a marker here.  If app calls is empty again the marker will be oldest and we can get a valid read on the status.
